src/AutoRest.CSharp/DataPlane/AutoRest/DataPlaneOutputLibrary.cs

```diff
@@ -207,17 +211,70 @@ public DataPlaneRestClient FindRestClient(OperationGroup operationGroup)
             return restClients;
         }
 
-        private static IEnumerable<LowLevelClientMethod> GetProtocolMethods(OperationGroup operationGroup, RestClientBuilder restClientBuilder, BuildContext<DataPlaneOutputLibrary> context)
+        // Get a Dictionary<operationGroupName, List<methodNames>> based on the "protocol-method-list" config
+        private Dictionary<string, List<string>> GetProtocolMethodsDictionary()
         {
-            // At least one protocol methods is found in the config
-            if (operationGroup.Operations.Where(IsProtocolMethodExists).Any())
+            Dictionary<string, List<string>> protocolMethodsDictionary = new();
+            foreach (var operationId in Configuration.ProtocolMethodList)
+            {
+                var operationGroupKeyAndIdArr = operationId.Split('_');
+
+                // If "operationGroup_operationId" passed in the config
+                if (operationGroupKeyAndIdArr.Length > 1)
+                {
+                    var operationGroupKey = operationGroupKeyAndIdArr[0];
+                    var methodName = operationGroupKeyAndIdArr[1];
+                    AddToProtocolMethodsDictionary(protocolMethodsDictionary, operationGroupKey, methodName);
+                }
+                // If operationGroup is not present, only operationId is passed in the config
+                else
+                {
+                    // when operationGroup is not present use the default name which is used to create client name
+                    var defaultClientName = ClientBuilder.GetClientPrefix(string.Empty, _context);
+                    AddToProtocolMethodsDictionary(protocolMethodsDictionary, defaultClientName, operationId);
+                }
+            }
+
+            return protocolMethodsDictionary;
+        }
+
+        private void AddToProtocolMethodsDictionary(Dictionary<string, List<string>> protocolMethodsDictionary, string operationGroupKey, string methodName)
+        {
+            if (!protocolMethodsDictionary.ContainsKey(operationGroupKey))
+            {
+                List<string> methodList = new();
+                methodList.Add(methodName);
+                protocolMethodsDictionary.Add(operationGroupKey, methodList);
+            }
+            else
+            {
+                var methodList = protocolMethodsDictionary[operationGroupKey];
+                methodList.Add(methodName);
+            }
+        }
+
+        private IEnumerable<string>? GetProtocolMethodsByOperationGroup(OperationGroup operationGroup)
+        {
+            _protocolMethodsDictionary.TryGetValue(operationGroup.Key, out var methodList);
+            if (methodList == null)
+            {
+                var defaultClientName = ClientBuilder.GetClientPrefix(string.Empty, _context);
```

Comment:
Why are we doing this fallback?

Reply:
If we pass only operationId without operationGroup in `protocol-method-list` like this case:
```
protocol-method-list:
  - Create
  ```
then we use default name by calling [ClientBuilder.GetClientPrefix()](https://github.com/Azure/autorest.csharp/blob/66398eee793fc828588cd56079af2c42abcb5704/src/AutoRest.CSharp/DataPlane/AutoRest/DataPlaneOutputLibrary.cs#L233) method for operationGroup key. I used the same logic as create client name for the methods where there is no operation group present with the operationId.

This way I get the correct client name where the method belongs.

Reply:
Correct me if I'm wrong, but wouldn't that mean that `Create` will be used in every client?

Reply:
No, Create method will not be generated in every client if we pass only `Create` in the config. Create method will be generated in the default RestClient.

Methods without operation groups will be generated in the default RestClient(in case when we pass only operationId in the config ex. `Create`) and methods with operation group will be generated in the configured operation group RestClient(in case when we pass operationGroup_operationId in the config ex. `firstTemplate_Create`, this will generate method in FirstTemplateRestClient)

Reply:
Can you add a test for it? E.g., add `secondTemplate_Create` operation to `ProtocolMethodsInRestClient.json` and validate that protocol method isn't generated for it.

Reply:
Already added all the possible scenarios in the test.

[Here](https://github.com/ShivangiReja/autorest.csharp/blob/protocol-method-list/test/TestProjects/ProtocolMethodsInRestClient/readme.md) is config, [here](https://github.com/ShivangiReja/autorest.csharp/tree/protocol-method-list/test/TestProjects/ProtocolMethodsInRestClient/Generated) is generated code and [here](https://github.com/ShivangiReja/autorest.csharp/blob/protocol-method-list/test/AutoRest.TestServer.Tests/ProtocolMethodsInRestClientTests.cs#L14-L56) are tests that validates where the methods are generated.

As an example you can see that `Create` method is not generated as protocol method in SecondTemplateRestClient.